### PR TITLE
payloadCreator arg argument => asyncThunk arg type

### PR DESCRIPTION
**Pull request by phryneas**
This would address #489 without changing any external signatures to this point. So @stephen-dahl and @msutkowski  please take a look.

The complete behavioral change should be covered by this test:

https://github.com/phryneas/redux-toolkit/blob/47eac023cd18e86cf888224381a77b9776a97677/type-tests/files/createAsyncThunk.typetest.ts#L185-L230

My only concern is the handling of optional parameters. For people with `strictNullChecks: false` in their `tsconfig.json`, this would make the parameter optional for all their asyncThunks - which is only consequent, but may be undesirable.
I've gotta take a second look if this can be done better - on the other hand, we have the same behavior with `createSlice` actions already, so I would not try too much there.

Also, this *might* make extending createAsyncThunk with Typescript (#486) a little more difficult (we'd have to do some things for that anyways, but it might be simpler before this change), so I've got to evaluate that before we merge it.

But please already take a look.

### CODE REVIEW

**Comment by msutkowski** [comment]
@phryneas This looks pretty good! Would it be a bad idea to change `ActionCreator` to the type below _(just swapping the last two lines)_? I believe this would address the issue you mentioned with parameters being optional - but I'm not 100% sure 😅 

```ts
type ActionCreator = IsAny<
    ThunkArg,
    (arg: ThunkArg) => AsyncActionThunkAction,
    unknown extends ThunkArg
      ? (arg: ThunkArg) => AsyncActionThunkAction
      : [ThunkArg] extends [void] | [undefined]
      ? () => AsyncActionThunkAction
      : [undefined] extends [ThunkArg] | []
      ? (arg: ThunkArg) => AsyncActionThunkAction
      : (arg?: ThunkArg) => AsyncActionThunkAction
  >
```

I'm also not sure at all about how this would impact the issue you referenced with extending the types for `createAsyncThunk`, but I can help test that tomorrow. 

My 🧠 has trouble processing those type tests this late at night, so I made [a gist of the common typing patterns](https://gist.github.com/msutkowski/81a69be15e96028c848bf3764334a323) to check against as well. (I just dropped this in a `recipes` folder locally against your branch - if you have a better way of doing something similar for these purposes, please let me know).

If you look at the last sample in that gist, I did `arg: string | void`, and I don't feel like that's too awful to type to mimic `arg?: string`.

Thanks again for putting this up, and sorry in advance if this isn't helpful!

**Comment by phryneas** [comment]
> ```ts
> type ActionCreator = IsAny<
>     ThunkArg,
>     (arg: ThunkArg) => AsyncActionThunkAction,
>     unknown extends ThunkArg
>       ? (arg: ThunkArg) => AsyncActionThunkAction
>       : [ThunkArg] extends [void] | [undefined]
>       ? () => AsyncActionThunkAction
>       : [undefined] extends [ThunkArg] | []
>       ? (arg: ThunkArg) => AsyncActionThunkAction
>       : (arg?: ThunkArg) => AsyncActionThunkAction
>   >
> ```

That would not prevent the issue, but introduce two new ones:

* a asyncThunk resulting from a payloadGenerator like `(arg?: number) => 0` would look like `(arg: number) => ...`
* a asyncThunk resulting from a payloadGenerator like `(arg: number) => 0` would look like `(arg?: number) => ...`

Essentially, the question is if we handle optional arguments or just replace the last three lines with `(arg: ThunkArg) => AsyncActionThunkAction`, making the argument mandatory in all cases where it *could* be different than `undefined`.

Something like `number | void` leading to `(arg?: number) => AsyncActionThunkAction` would be a possibility, too, of course, but that would not be very intuitive so I'm not sure if it would even be worth it to have it there.

Gotta run to work now :)

**Comment by stephen-dahl** [comment]
> Something like `number | void` leading to `(arg?: number) => AsyncActionThunkAction` would be a possibility, too, of course, but that would not be very intuitive so I'm not sure if it would even be worth it to have it there.

would there be another way to handle `(arg: number | void, thunkApi) => {}`?
`arg` cant be `arg?` in this case we cant have an optional param before a required one.

figuring out types for some of these edge cases looks like a pain 😂

**Comment by msutkowski** [comment]
@phryneas Gotcha. In that case, the only issue I see with the current types is this:

```ts
const usingInline = createAsyncThunk(
  'test/usingInline',
  (arg: string, { getState }) => {
    const { test } = getState() as RootState

    return {
      banana: 'test'
    }
  }
)

usingInline() // i would expect an error being that we say `arg: string`, but this doesn't get caught as the resulting type is `arg?: string`
```

If there is a way to get `arg: string` and `arg?: string` to work as expected, that'd be the ideal thing. If not, I think I would _personally_ prefer the `string | void` to imply an optional argument. The reason being is the typing can be weird. Consider these:

```ts
const usingInline = createAsyncThunk(
  'test/usingInline',
  (arg?: string, { getState }) => { // results in - A required parameter cannot follow an optional parameter.
    const { test } = getState() as RootState

    return {
      banana: 'test'
    }
  }
)
```

```ts
// you can't do `string?` here as the `arg` param, so how would a user accomplish this?
const usingInline = createAsyncThunk<{ banana: string }, string, { state: RootState }>(
  'test/usingInline',
  (arg, { getState }) => {
    const { test } = getState()

    return {
      banana: 'test'
    }
  }
)
```

My only other input is that I don't think `arg` will be typed as optional by a user very often. It seems it is either well-defined or void. I could be wrong here, but that's what I've seen so far in Discord. Thanks!

**Comment by phryneas** [comment]
I've added lots of tests for usage with a seconds argument.

The "optional" cases will now be optional with `strictNullChecks: true`, but require `| void` declaration with `strictNullChecks: false` so not all arguments become optional. I believe this is the best compromise we can get.

Have I missed any weird case?

File: type-tests/files/createAsyncThunk.typetest.ts [hunk 1]
```diff
@@ -181,3 +181,164 @@ const defaultDispatch = (() => {}) as ThunkDispatch<{}, any, AnyAction>
...
+    expectType<(arg?: number) => any>(asyncThunk)
+    asyncThunk()
+    asyncThunk(5)
+    // typings:expect-error
+    asyncThunk('string')
+  }
+
+  // two arguments, first specified as any: asyncThunk has required any argument
+  {
+    const asyncThunk = createAsyncThunk('test', (arg: any) => 0)
```
Comment (stephen-dahl) [inline type-tests/files/createAsyncThunk.typetest.ts:320]: shouldnt this be
```suggestion
    const asyncThunk = createAsyncThunk('test', (arg: any, thunkApi) => 0)
```

Comment (phryneas) [reply type-tests/files/createAsyncThunk.typetest.ts]: You're absolutely right... :facepalm:

**Comment by phryneas** [comment]
Thanks... guess I was tired. Added those, I hopen that were all.

**Comment by stephen-dahl** [comment]
thanks @phryneas I think this is the best experience I have had with getting something updated in a project. lgtm

**Comment by phryneas** [comment]
@stephen-dahl  I still have to extract some helper types to make #486 less painful, but now with the tests in place & thus the behaviour reviewed from you it's gonna be pretty straightforward - I'll try to finish this up today or tomorrow :)

**Comment by msutkowski** [comment]
@phryneas Once #486 is addressed, I'll open a PR for the docs with [common typing recipes](https://gist.github.com/msutkowski/81a69be15e96028c848bf3764334a323), and include an example of typing a generic helper per that thread.

**Comment by phryneas** [comment]
Okay, I've just extracted these types out and exported one of them, `AsyncThunkAction` - that should help make things more readable in tooltips and error messages.

When the tests are finished, I believe this is ready to merge from my side. @markerikson what do you think?

File: src/createAsyncThunk.ts [hunk 2]
```diff
@@ -103,6 +103,64 @@ type GetRejectValue<ThunkApiConfig> = ThunkApiConfig extends {
...
+  dispatch: GetDispatch<ThunkApiConfig>,
+  getState: () => GetState<ThunkApiConfig>,
+  extra: GetExtra<ThunkApiConfig>
+) => Promise<
+  AsyncThunkReturnValue<ThunkArg, Returned, GetRejectValue<ThunkApiConfig>>
+> & {
+  abort(reason?: string): void
+}
+
+type AsyncThunkActionCreator<
```
Comment (markerikson) [inline src/createAsyncThunk.ts:137]: I sure hope you know what you're doing here, because my eyes are glazing over

**Comment by markerikson** [comment]
Rebased on master after merging #512 .

File: src/createAsyncThunk.ts (hunk 2 above)
Comment (phryneas) [reply src/createAsyncThunk.ts:164]: I added more tests than I added code, guess we'll be fine :)
